src/integer_mod_q/polynomial_ring_zq/set.rs

```diff
@@ -0,0 +1,171 @@
+    /// use qfall_math::integer::PolyOverZ;
+    /// use qfall_math::integer_mod_q::{PolynomialRingZq, ModulusPolynomialRingZq};
+    /// use std::str::FromStr;
+    ///
+    /// let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
+    /// let poly = PolyOverZ::from_str("3  0 1 1").unwrap();
+    /// let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+    ///
+    /// poly_ring.set_coeff(2, 16).unwrap();
+    /// ```
+    ///
+    /// # Errors and Failures
+    /// - Returns a [`MathError`] of type [`OutOfBounds`](MathError::OutOfBounds) if
+    /// either the index is negative or it does not fit into an [`i64`].
+    fn set_coeff(
+        &mut self,
+        index: impl TryInto<i64> + Display,
+        value: Integer,
+    ) -> Result<(), MathError> {
+        let value = value.into();
+        let index = evaluate_index(index)?;
+
+        unsafe {
+            fmpz_poly_set_coeff_fmpz(&mut self.poly.poly, index, &value.value);
+        };
+        self.reduce();
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test_set_coeff {
+    use crate::{
+        integer::{PolyOverZ, Z},
+        integer_mod_q::{ModulusPolynomialRingZq, PolynomialRingZq},
+        traits::SetCoefficient,
+    };
+    use std::str::FromStr;
+
+    /// Ensure that setting is available for other types.
+    #[test]
+    fn availability() {
+        let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
+        let poly = PolyOverZ::from_str("3  0 1 1").unwrap();
+        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+
+        poly_ring.set_coeff(1, 3u64).unwrap();
+        poly_ring.set_coeff(1, 3u32).unwrap();
+        poly_ring.set_coeff(1, 3u16).unwrap();
+        poly_ring.set_coeff(1, 3u8).unwrap();
+        poly_ring.set_coeff(1, 3i64).unwrap();
+        poly_ring.set_coeff(1, 3i32).unwrap();
+        poly_ring.set_coeff(1, 3i16).unwrap();
+        poly_ring.set_coeff(1, 3i8).unwrap();
+        poly_ring.set_coeff(1, Z::from(3)).unwrap();
+    }
+
+    /// Ensure that coefficients up to 2^15 -1 work.
+    #[test]
+    fn set_max_coeff() {
+        let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
+        let poly = PolyOverZ::from_str("3  0 1 1").unwrap();
+        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+
+        assert!(poly_ring.set_coeff(i8::MAX, 2).is_ok());
+        assert!(poly_ring.set_coeff(i16::MAX, 2).is_ok());
+    }
+
+    /// Ensure that the max of [`u8`] and [`u16`] works as a coefficient.
+    #[test]
+    fn set_unsigned_coeff() {
+        let modulus = ModulusPolynomialRingZq::from_str("4  1 0 0 1 mod 17").unwrap();
+        let poly = PolyOverZ::from_str("3  0 1 1").unwrap();
+        let mut poly_ring = PolynomialRingZq::from((&poly, &modulus));
+
+        assert!(poly_ring.set_coeff(u8::MAX, 2).is_ok());
+        assert!(poly_ring.set_coeff(u16::MAX, 2).is_ok());
+    }
```

Comment:
I think that you can remove `set_max_coeff`. It is basically the same test as `set_unsigned_coeff`, but with a smaller index.

Additionally, the wording should be clearified to more clearly differentiate between `index` and `value`. I thought of value when I read the word coefficient, but the u16::MAX th coefficient is set.

Furthermore, I think you can simplify the test by setting the poly to `PolyOverZ::default()`.